firebase-app-distribution/src/main/java/com/google/firebase/app/distribution/FirebaseAppDistribution.java

```diff
@@ -145,28 +151,117 @@ public UpdateTask updateIfNewReleaseAvailable() {
                 setCachedUpdateIfNewReleaseResult();
                 return Tasks.forResult(null);
               }
-              return showUpdateAlertDialog(release);
-            })
-        .addOnFailureListener(
-            e -> {
-              postProgressToCachedUpdateIfNewReleaseTask(
-                  UpdateProgress.builder()
-                      .setApkFileTotalBytes(UNKNOWN_RELEASE_FILE_SIZE)
-                      .setApkBytesDownloaded(UNKNOWN_RELEASE_FILE_SIZE)
-                      .setUpdateStatus(UpdateStatus.NEW_RELEASE_CHECK_FAILED)
-                      .build());
-              setCachedUpdateIfNewReleaseCompletionError(
-                  e,
-                  new FirebaseAppDistributionException(
-                      Constants.ErrorMessages.NETWORK_ERROR,
-                      FirebaseAppDistributionException.Status.NETWORK_FAILURE));
-            });
+              return showUpdateAlertDialog(release);}
+              )
+        .onSuccessTask( unused ->
+            updateApp(true)
+                .addOnProgressListener(
+                    this::postProgressToCachedUpdateIfNewReleaseTask)
+                .addOnFailureListener(
+                    this::setCachedUpdateIfNewReleaseCompletionError))
+        .addOnFailureListener(this::setCachedUpdateIfNewReleaseCompletionError);
 
     synchronized (updateIfNewReleaseTaskLock) {
       return cachedUpdateIfNewReleaseTask;
     }
   }
 
+  private Task<Void> showSignInDialog() {
+    if (isTesterSignedIn()) {
+      return Tasks.forResult(null);
+    }
+
+    TaskCompletionSource<Void> showDialogTask = new TaskCompletionSource<>();
+
+    Activity currentActivity = lifecycleNotifier.getCurrentActivity();
+    if (currentActivity == null) {
+      LogWrapper.getInstance().e("No foreground activity found.");
+      showDialogTask.setException(
+          new FirebaseAppDistributionException(
+              ErrorMessages.APP_BACKGROUNDED,
+              FirebaseAppDistributionException.Status.UPDATE_NOT_AVAILABLE));
```

Comment:
Seems like there are scenarios like this where the status doesn't exactly align with what is going on (here its not that an update isn't available). Our logs do have the more specific case to give the user a better sense of whats going on, but do we want to look at these statuses to make them more specific?

Reply:
This is a really good call. I've added an error message enum to our API and made a note to add it to the API addendum.

Reply:
Agreed, good callout. There are a bunch of places like this where the error status and/or message does not align with the actual situation. I'm currently going through them all as part of the error handling cleanup, so @rachaprince if you want I can handle this in a followup since this was just moved code.

For this case though, looking at the statuses, `AUTHENTICATION_FAILED` might work if we consider "authentication" to be synonymous with "tester sign in."

Reply:
Actually I think your new "no foreground activity" error seems like a good one to add, could be useful elsewhere. `AUTHENTICATION_FAILED` probably implies that the actual authentication was not allowed, which would be misleading.